### PR TITLE
Add authorization support based on allowed actions

### PR DESCRIPTION
**Pull request by zhang-hua**
* Get the user role info from JWT token
* Query authentication service to get allowed actions defined in role policies
* Enforce the policy of allowed actions

# Type of change? 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

...

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -91,6 +94,7 @@ more information. More information on environment variables
 
 1. Make sure the [Prerequisites](#prerequisites) are set up.
 1. Set the following environment variables in your system. More information on environment variables [here](#configuration-and-environment-variables).
+	1. `PCS_AUTH_WEBSERVICE_URL` = {authentication service endpoint}
```
Comment (ppathan) [inline README.md]: nit: Extra space

File: WebService/Auth/AuthMiddleware.cs [hunk 2]
```diff
@@ -181,6 +188,20 @@ private bool ValidateToken(string token, HttpContext context)
...
+                    var userObjectId = context.Request.GetCurrentUserObjectId();
+                    var roles = context.Request.GetCurrentUserRoleClaim().ToList();
+                    if (roles.Any())
+                    {
+                        var allowedActions = this.userManagementClient.GetAllowedActionsAsync(userObjectId, roles).Result;
+                        context.Request.SetCurrentUserAllowedActions(allowedActions);
+                    }
+                    else
+                    {
+                        this.log.Error("JWT token doesn't include any role claims.", () => { });
```
Comment (ppathan) [inline WebService/Auth/AuthMiddleware.cs]: Is this an error? If it is then should we still return true or false?

File: WebService/Auth/RequestExtension.cs [hunk 3]
```diff
@@ -1,30 +1,88 @@
...
+            if (!request.HttpContext.Items.ContainsKey(CONTEXT_KEY_USER_CLAIMS))
             {
                 return new List<Claim>();
             }
 
-            return request.HttpContext.Items[CONTEXT_KEY] as IEnumerable<Claim>;
+            return request.HttpContext.Items[CONTEXT_KEY_USER_CLAIMS] as IEnumerable<Claim>;
+        }
+
+
```
Comment (ppathan) [inline WebService/Auth/RequestExtension.cs]: nit: extra line

File: WebService/Auth/AuthMiddleware.cs (hunk 2 above)
Comment (zhang-hua) [reply WebService/Auth/AuthMiddleware.cs]: if no role claim in the valid token, then request can still be processed as 'ReadOnly' role by default and doesn't need to be failed.

File: scripts/docker/run [hunk 4]
```diff
@@ -15,6 +15,7 @@ run_container() {
 	docker run -it -p 9002:9002 \
         -e PCS_IOTHUB_CONNSTRING \
         -e PCS_STORAGEADAPTER_WEBSERVICE_URL \
+        -e PCS_AUTH_WEBSERVICE_URL \
```
[the remark this answers is not in the document]
Comment (zhang-hua) [reply scripts/docker/run]: no, I think.

File: README.md [hunk 5]
```diff
@@ -64,6 +64,8 @@ at least once. See specific instructions for IDE or command line setup below for
 more information. More information on environment variables
 [here](#configuration-and-environment-variables).
 
+* `PCS_AUTH_WEBSERVICE_URL` = http://localhost:9001/v1
+    * The url for the [Authentication service](https://github.com/Azure/pcs-auth-dotnet) from [Setup Dependencies](#setup-dependencies)
```
Comment (jillcary) [inline README.md]: Can you add the auth service to setup dependencies section?

File: WebService/Runtime/Config.cs [hunk 6]
```diff
@@ -76,6 +79,7 @@ public Config(IConfigData configData)
                 DevicePropertiesTTL = configData.GetInt(DEVICE_PROPERTIES_TTL_KEY),
                 DevicePropertiesRebuildTimeout = configData.GetInt(DEVICE_PROPERTIES_REBUILD_TIMEOUT_KEY),
                 StorageAdapterApiUrl = configData.GetString(STORAGE_ADAPTER_URL_KEY),
+                UserManagementApiUrl = configData.GetString(USER_MANAGEMENT_URL_KEY),
```
Comment (jillcary) [inline WebService/Runtime/Config.cs]: nit, remove comma

**Pull request merged**